### PR TITLE
WIP: WSIViewer

### PR DESCRIPTION
**Pull request by aacic**
## Description

Adds whole slide image viewer to the project.

The client-side code is extended by the open-layers library and ol - extension library:

https://openlayers.org/en/latest/examples/zoomify.html

https://viglino.github.io/ol-ext/examples/control/map.switcher.image.html

The map switcher is not the perfect solution, since UI/UX is not perfect for this use case, and also there is an issue with css esbuild bundling and I had to implement a custom escape fix:
`client/postcss-escape-fix.mjs `

As a next step, we could implement an image switcher from scratch.  

Zoomify instance displays `WSIs` and `Layerswitcher` to switch between different images.

The server-side code is extended by `tough-cookie` library which is used to read and use session cookies for `TileServer` requests. 

`TileServer` is a new component introduced to the project as a docker image, under:

`proteinpaint/container/tileserver`

The component consists of the `TileServer` app from the `TIAToolbox` project. 

https://tia-toolbox.readthedocs.io/en/latest/_autosummary/tiatoolbox.visualization.tileserver.TileServer.html#tiatoolbox.visualization.tileserver.TileServer

To test:

Check out `add-wsimages-to-nintendo-public ` from sjpp folder.

Extend the`nintendo-public` dataset by adding the folder:

`~/data/tp/mm10/nintendo-public/wsimages/B-T87L964D`

Inside of the folder, place the .svs image:
https://portal.gdc.cancer.gov/files/73c69d24-6f9e-44e2-bfe5-a608d4cf5c27

Run `TileServer`:

cd /container/tileserver
./run.sh "/path/to/tp/"

Extend `serverconfig.json` by adding:

```
"features": {
   "showWSImages": 1,
},
```

```
   "tileServerURL": "http://localhost:5000",
   "tileServerMount": "/home/root/tileserver/tp/"
```

Open:
http://localhost:3000/?noheader=1&mass={%22genome%22:%22mm10%22,%22dslabel%22:%22nintendoPublic%22}

## Checklist

[Check each task](https://github.com/stjude/proteinpaint/wiki/Pull-Request-Checklist) that has been performed or verified to be not applicable.
- [x] Tests: added and/or passed unit and integration tests, or N/A
- [x] Todos: commented or documented, or N/A
- [x] Notable Changes: updated release.txt, prefixed a commit message with "fix:" or "feat:", added to an internal tracking document, or N/A

### CODE REVIEW

**Review by siosonel — COMMENTED** [review]
I only reviewed the css quick fixes, looks good for now, can improve later.

File: server/routes/samplewsimages.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+			res.send({ sampleWSImages: sampleWSImages })
+		} catch (e: any) {
+			console.log(e)
+			res.status(404).send('Sample images not found')
+		}
+	}
+}
+
+function getWSImages(sampleImagesPath: string): string[] {
+	const files = fs.readdirSync(sampleImagesPath)
```
Comment (xzhou82) [inline server/routes/samplewsimages.ts:52]: replace with fs.promises.readdir()

File: server/shared/types/dataset.ts [hunk 2]
```diff
@@ -598,6 +598,7 @@ type Mds3Queries = {
 	singleSampleGenomeQuantification?: SingleSampleGenomeQuantification
 	singleSampleGbtk?: SingleSampleGbtk
 	DZImages?: any
+	WSImages?: any
```
Comment (xzhou82) [inline server/shared/types/dataset.ts:601]: should declare obj as {imageBySampleFolder, type} and clarify purpose of each

File: server/routes/termdb.config.ts [hunk 3]
```diff
@@ -219,6 +219,12 @@ function addNonDictionaryQueries(c, ds: Mds3WithCohort, genome) {
 		}
 	}
 
+	if (q.WSImages) {
+		q2.WSImages = {
```
Comment (xzhou82) [inline server/routes/termdb.config.ts:223]: add the feature toggle here

File: server/shared/types/dataset.ts (hunk 2 above)
Comment (aacic) [reply server/shared/types/dataset.ts]: Fixed.

File: server/routes/termdb.config.ts (hunk 3 above)
Comment (aacic) [reply server/routes/termdb.config.ts:223]: Fixed.

File: server/routes/samplewsimages.ts (hunk 1 above)
Comment (aacic) [reply server/routes/samplewsimages.ts]: Fixed.

**Review by airenzp — APPROVED** [review]
I could reproduce it and it works. The code looks good overall. Better to use the nintendo image instead. I am approving. Thanks!

**Comment by xzhou82** [comment]
http://localhost:3000/testrun.html breaks backend, please check

**Review by xzhou82 — APPROVED** [review]

**Pull request merged**